src/arithmetic.jl

```diff
@@ -250,93 +248,85 @@ for T in (:HomogeneousPolynomial, :TaylorN)
     end
 end
 
+for (T, W) in ((:Taylor1, :Number), (:TaylorN, :NumberNotSeriesN))
+    @eval *{T<:$W, S<:$W}(a::$T{T}, b::$T{S}) = *(promote(a,b)...)
 
+    @eval function *{T<:$W}(a::$T{T}, b::$T{T})
+        corder = max(a.order, b.order)
+        if $T == Taylor1
+            c = Taylor1(zero(a[1]), corder)
+        else
+            c = TaylorN(zeros(HomogeneousPolynomial{T}, corder))
+        end
 
-doc"""
-```
-*(a, b)
-```
-
-Return the Taylor expansion of $a \cdot b$, of order `max(a.order,b.order)`, for
-`a::Taylor1`, `b::Taylor1` polynomials.
-
-For details on making the Taylor expansion, see [`TaylorSeries.mulHomogCoef`](@ref).
-"""
-*{T<:Number,S<:Number}(a::Taylor1{T}, b::Taylor1{S}) = *(promote(a,b)...)
-
-function *{T<:Number}(a::Taylor1{T}, b::Taylor1{T})
-    a, b = fixorder(a, b)
-    coeffs = similar(a.coeffs)
-    @inbounds for k = 0:a.order
-        coeffs[k+1] = mulHomogCoef(k, a.coeffs, b.coeffs)
+        @inbounds for ord = 0:corder
+            mul!(c, a, b, ord) # updates c[ord+1]
+        end
+        return c
     end
-    return Taylor1(coeffs, a.order)
 end
 
-*{T<:NumberNotSeriesN,S<:NumberNotSeriesN}(a::TaylorN{T}, b::TaylorN{S}) = *(promote(a,b)...)
 
-function *{T<:NumberNotSeriesN}(a::TaylorN{T}, b::TaylorN{T})
-    a, b = fixorder(a, b)
-    coeffs = zeros(HomogeneousPolynomial{T}, a.order)
+*{T<:NumberNotSeriesN,S<:NumberNotSeriesN}(a::HomogeneousPolynomial{T},
+    b::HomogeneousPolynomial{S}) = *(promote(a,b)...)
 
-    for ord in eachindex(coeffs)
-        for i = 0:ord-1
-            @inbounds mul!(coeffs[ord], a[i+1], b[ord-i])
-        end
-    end
-
-    return TaylorN(coeffs, a.order)
-end
+function *{T<:NumberNotSeriesN}(a::HomogeneousPolynomial{T},
+        b::HomogeneousPolynomial{T})
 
-## Multiplication ##
-*{T<:NumberNotSeriesN,S<:NumberNotSeriesN}(a::HomogeneousPolynomial{T}, b::HomogeneousPolynomial{S}) =
-    *(promote(a,b)...)
-
-function *{T<:NumberNotSeriesN}(a::HomogeneousPolynomial{T}, b::HomogeneousPolynomial{T})
     order = a.order + b.order
 
-    order > get_order() && return HomogeneousPolynomial([zero(a[1])], get_order())
+    order > get_order() &&
+        return HomogeneousPolynomial([zero(a[1])], get_order())
```

Comment:
This line and the next can be combined with something like  `min(order, get_order())`.
Isn't there a better way of writing this zero?